tests/ChannelIdempotentTest.php

```diff
@@ -0,0 +1,128 @@
+
+
+class ChannelIdempotentTest extends \PHPUnit_Framework_TestCase {
+    protected static $testApp;
+    protected static $defaultOptions;
+    protected static $ably;
+
+    public static function setUpBeforeClass() {
+        self::$testApp = new TestApp();
+        self::$defaultOptions = self::$testApp->getOptions();
+        self::$ably = new AblyRest( array_merge( self::$defaultOptions, [
+            'key' => self::$testApp->getAppKeyDefault()->string,
+            'idempotentRestPublishing' => true,
+        ] ) );
+    }
+
+    public static function tearDownAfterClass() {
+        self::$testApp->release();
+    }
+
+    /**
+     * RSL1j
+     */
+    public function testMessageSerialization() {
+        $channel = self::$ably->channel( 'messageSerialization' );
+
+        $msg = new Message();
+        $msg->name = 'name';
+        $msg->data = 'data';
+        $msg->clientId = 'clientId';
+        $msg->id = 'foobar';
+
+        $body = $channel->__publish_request_body( $msg );
+        $body = json_decode($body);
+
+        $this->assertTrue( property_exists($body, 'name') );
+        $this->assertTrue( property_exists($body, 'data') );
+        $this->assertTrue( property_exists($body, 'clientId') );
+        $this->assertTrue( property_exists($body, 'id') );
+    }
+
+    /**
+     * RSL1k1
+     */
+    public function testIdempotentLibraryGenerated() {
+        $channel = self::$ably->channel( 'idempotentLibraryGenerated' );
+
+        $msg = new Message();
+        $msg->name = 'name';
+        $msg->data = 'data';
+
+        $body = $channel->__publish_request_body( $msg );
+        $body = json_decode($body);
+
+        $id = explode ( ":", $body->id);
+        $this->assertEquals( count($id), 2);
+        $this->assertGreaterThanOrEqual( base64_decode($id[0]), 9);
+        $this->assertEquals( $id[1], "0");
+    }
+
+    /**
+     * RSL1k2
+     */
+    public function testIdempotentClientSupplied() {
+        $channel = self::$ably->channel( 'idempotentClientSupplied' );
+
+        $msg = new Message();
+        $msg->name = 'name';
+        $msg->data = 'data';
+        $msg->id = 'foobar';
+
+        $body = $channel->__publish_request_body( $msg );
+        $body = json_decode($body);
+        $this->assertEquals( $body->id, "foobar" );
+    }
+
+    /**
+     * RSL1k3
+     */
+    public function testIdempotentMixedIds() {
+        $channel = self::$ably->channel( 'idempotentMixedIds' );
+
+        $messages = [];
+
+        $msg = new Message();
+        $msg->name = 'name';
+        $msg->data = 'data';
+        $msg->id = 'foobar';
+        $messages[] = $msg;
+
+        $msg = new Message();
+        $msg->name = 'name';
+        $msg->data = 'data';
+        $messages[] = $msg;
+
+        $body = $channel->__publish_request_body( $messages );
+        $body = json_decode($body);
+
+        $this->assertEquals( $body[0]->id, "foobar" );
+        $this->assertFalse( property_exists($body[1], 'id') );
```

Comment:
This should fail, you can't mix messages with idempotency and not idempotency, so not sure why this works.

Reply:
Don't understand. RSL1k3 says `If more than one Message is passed to publish() and one or more of those messages contains a non-empty id attribute, then all message ids (present or absent) are preserved on sending the batch of messages.`

That's what we're testing. The first message with `id`, the second without. In line 105 we test the id is sent for th 1st message, in line 106 we test it's not sent for the 2nd message.

Reply:
Just run the tests locally and they still pass, so I guess the quick fix has not been implemented?

Reply:
@paddybyers ^?

Reply:
I've just written a test that sends an identical 2 messages and they get rejected. @jdavid are you testing this against the `idempotent-dev` environment?

Reply:
@paddybyers yes, in the `idempotent` branch, since commit 2e04a82af8beda72c170d8da420c6676d478c23d

Reply:
@jdavid ok then, how easy is it for me to get set up to run the tests myself to see what's going on?